keras_spiking/model_energy.py

```diff
@@ -0,0 +1,654 @@
+        )
+
+    def __init__(self, model, input_data=None, timesteps=1):
+        self.layer_stats = {}
+        self.layer_rates = {}
+        self.node_stats = {}
+
+        self.model = model
+        self.input_data = input_data
+        self.timesteps = timesteps
+
+        if timesteps != 1 and any(
+            x.shape[1] != self.timesteps for x in self.model.inputs
+        ):
+            warnings.warn(
+                "`timesteps` (%s) should match the number of timesteps in the model "
+                "input(s) [%s]"
+                % (
+                    self.timesteps,
+                    ", ".join(str(x.shape[1]) for x in self.model.inputs),
+                )
+            )
+
+        self._compute_model_stats()
+        if self.input_data is not None:
+            self._compute_model_rates(self.input_data)
+
+    def _compute_model_stats(self):
+        """Compute statistics for ``self.model``."""
+        for nodes in self.model._nodes_by_depth.values():
+            for node in nodes:
+                self._compute_node_stats(node)
+
+    def _compute_node_stats(self, node):
+        """Compute statistics for the given node."""
+        if node in self.node_stats:
+            return  # pragma: no cover
+
+        layer = node.layer
+        self.node_stats[node] = self.compute_layer_stats(
+            layer, timesteps=self.timesteps, node=node
+        )
+        self.layer_stats.setdefault(layer, {})
+        add_dict_to_dict(self.layer_stats[layer], self.node_stats[node])
+
+    def _compute_model_rates(self, input_data):
+        """Compute layer input firing rates for the given data."""
+
+        relevant_nodes = [
+            node for nodes in self.model._nodes_by_depth.values() for node in nodes
+        ]
+
+        layer_inputs = []
+        for layer in self.model.layers:
+            inp_tensors = []
+            for in_node in layer.inbound_nodes:
+                if in_node in relevant_nodes:
+                    inp_tensors.append(in_node.input_tensors)
+
+            layer_inputs.append(inp_tensors)
+
+        layer_model = tf.keras.Model(self.model.inputs, layer_inputs)
+        rates = layer_model.predict(input_data)
+        assert len(rates) == len(layer_inputs)
+
+        for layer, rates_in in zip(self.model.layers, rates):
+            self.layer_rates[layer] = np.sum(
+                [tf.math.reduce_mean(x).numpy() for x in rates_in]
+            )
+
+    def _layer_energy(
+        self,
+        stats,
+        synop_energy,
+        neuron_energy,
+        timesteps,
+        rate=None,
+        dt=0.001,
+        spiking=True,
+    ):
+        """Compute the energy used by one layer"""
+        # assume neurons communicate once per timestep if no rate provided
+        rate = (1 / dt) if rate is None or not spiking else rate
+        timesteps = 1 if not spiking else timesteps
+        return dict(
+            synops=synop_energy * stats.get("synops", 0) * rate * timesteps * dt,
```

Comment:
Isn't this assuming that non-spiking values (e.g. model inputs, outputs of non-spiking layers) are spike-encoded? E.g., if a non-spiking layer (the input to this layer) is outputting a value of 100, this is treating that as 100 events per second (not a value of 100 output every timestep).

Reply:
I guess that's not a bad assumption (since spiking devices, like Loihi/Spinnaker, typically force the inputs to be spike encoded). So maybe we should just make that clear in the docs somewhere.